app/controllers/katello/api/v2/upstream_subscriptions_controller.rb

```diff
@@ -27,10 +27,30 @@ def index
       respond(collection: collection)
     end
 
+    api :POST, "/organizations/:organization_id/upstream_subscriptions",
+      N_("Add subscriptions consumed by a manifest from Red Hat Subscription Management")
+    param :organization_id, :number, desc: N_("Organization ID"), required: true
+    param :pools, Array, desc: N_("Array of pool IDs to add"), required: true do
+      param :id, String, desc: N_("Pool ID"), required: true
+      param :quantity, :number, desc: N_("Quantity of entitlements to bind")
+    end
+    def create
+      UpstreamConsumer.bind_entitlements(@organization, bind_entitlements_params)
+      render json: UpstreamConsumer.get_entitlements(@organization, cp_search_params)
```

Comment:
Not sure if returning all attached subscriptions after a bind is a good idea. this could be *really* slow for some large customers. I'd rather fetch that data only if we need it

Reply:
I also think when we groomed this and related cards we talked about using dynflow actions to manage things like add and delete - do you recall that? What do you think about using an action? I think it could be good for debugging purposes, but curious what you and others feel about it.

Reply:
Don't worry, the view isn't finished here. ;)

Reply:
I don't remember any talk of using foreman_tasks for this. It's not a long-running request or anything. I'm not opposed to it, but right now I'm not sure of the benefit.

Reply:
Yes, we need to use foreman-tasks, as the bind of entitlements isn't the whole matter, we have to also do a full manifest refresh after the bind

Reply:
and it should be an async task as well, returning the task structure similar to other async tasks

Reply:
I don't understand why we need to do a whole manifest refresh (I know that's what the plan is). We get back the new entitlement(s) from this request with pools embedded. We have all the information we need locally at that point. Isn't there a way to call RefreshRepository with our newly provided products?

Reply:
Candlepin doesn't have this new data though.  That's what the manifest refresh does.  We discussed just trying to update the pool quantities directly in candlepin, but they didn't like that idea.  (here though you're adding completely new pools, which i think should do a refresh regardless.)

Reply:
Which Candlepin doesn't have the new data? And by new data you mean the new entitlements?